### ISSUE TITLE
Styles are not applied to react-select inside a shadow-DOM component

### ISSUE THREAD

**Issue by pusulurm** [JedWatson/react-select#3680]
I'm using react-select in my application and it is been working great until I hit this situation. 
I have wrapped my react application as a shadow-DOM component to isolate styles from other applications. After this change react-select does not get any styles applied.
As per my analysis, react-select uses emotion library to build dynamic css and these styles are being added to document>head section. Since my application is a Shadow-DOM wrapped application, these external styles are not accessible to react-select.

I can see option to write my custom styles but that is very tidies to write all styles and upgrades will become unmanageable. Please suggest a solution for this case.

**Comment by pusulurm** [comment]
I've received a suggestion from emotions library group to use a custom container. 
https://github.com/emotion-js/emotion/issues/388

Could not achieve this. Any help on this is greatly appreciated.

**Comment by afvieira** [comment]
I have the same problem. The problem is related to "screen readers" feature.
When I open the menu, it adds a `<span>` above the dropdown (see the image below).
![image](https://user-images.githubusercontent.com/1200312/63850978-30ee3500-c98d-11e9-9951-98d0984e60c7.png)
Also, that `<span>` has a class name `css-bl6clz` that doesn't exist (in a shadow-DOM component context) and this results in a strange behavior:
![image](https://user-images.githubusercontent.com/1200312/63851665-6f382400-c98e-11e9-9a26-270ab1474d39.png)

By the way, is there any way to deactivate this feature ("screen readers"), in order to not render this kind of dynamically `html`?

**Comment by afvieira** [comment]
> I've received a suggestion from emotions library group to use a custom container.
> [emotion-js/emotion#388](https://github.com/emotion-js/emotion/issues/388)
> 
> Could not achieve this. Any help on this is greatly appreciated.

@pusulurm did you manage to solve this any other way?

**Comment by norama** [comment]
@afvieira @pusulurm I have managed to solve this with a little workaround:

```

**Comment by wenisy** [comment]
same question here, I used it in Iframe, but when I click selector, it would add styles in parent window head.

**Comment by pusulurm** [comment]
> > I've received a suggestion from emotions library group to use a custom container.
> > [emotion-js/emotion#388](https://github.com/emotion-js/emotion/issues/388)
> > Could not achieve this. Any help on this is greatly appreciated.
> 
> @pusulurm did you manage to solve this any other way?

@afvieira , As I was in urgency , I ended up downgrading react select to 1.x version. This comes with a css file and I could bundle with my main css file.
Since my bundle css is scoped under shadow DOM it solved my problem. But I had to loose all latest features of react-select.

@norama 
Thank you for sharing your solution. I'm glad you have some working solution.
I tried this approach and I have an issue and hope you can suggest on this.
With my own NonceProvider, I am able to get the style sections injected in my webcomponent and react select works fine with it. But I could see duplicate style sections added every time I render react select component. Where I have 10 select elements, the styles are duplicated 10 time.

Is there a way to avoid this flood of style tags from emotion ? Thanks again

**Comment by bonzoSPb** [comment]
> ```js
> class MyNonceProvider extends NonceProvider {
>     createEmotionCache = (nonce) => {
>         return createCache({ nonce, container: this.props.container });
>     };
> }
> ```

didnt work on mobile

**Comment by bonzoSPb** [comment]
@JedWatson 
we additional check and now it works
Can you update your package?
`js
    _this.onTouchEnd = function (event) {
      if (_this.userIsDragging) return; // close the menu if the user taps outside
      // we're checking on event.target here instead of event.currentTarget, because we want to assert information
      // on events on child elements, not the document (which we've attached this handler to).
 
      if ((_this.controlRef && !_this.controlRef.contains(event.target) && _this.menuListRef && !_this.menuListRef.contains(event.target)) &&
      (_this.controlRef && !_this.controlRef.contains(event.composedPath()[0]) && _this.menuListRef && !_this.menuListRef.contains(event.composedPath()[0]))) {
        _this.blurInput();
      } // reset move vars
 
 
      _this.initialTouchX = 0;
      _this.initialTouchY = 0;
    };`

**Comment by harman666666** [comment]
please update jet. thanks

**Comment by ErikAGriffin** [comment]
I'm having an issue where React-Select seems to not be working at all inside the shadow DOM.  Focusing the element, onChange, nothing seems to be working properly.  Everyone else's issue is just that the styles aren't being applied?

**Comment by n3bhmb** [comment]
@bonzoSPb 
Hi, where did you add that additionnal check ? In which file ?

**Comment by bonzoSPb** [comment]
> @bonzoSPb
> Hi, where did you add that additionnal check ? In which file ?
In Select.js, check my fork
https://github.com/bonzoSPb/react-select/blob/master/src/Select.js

**Comment by n3bhmb** [comment]
@bonzoSPb thank you 👍

**Comment by n3bhmb** [comment]
> > @bonzoSPb
> > Hi, where did you add that additionnal check ? In which file ?
> 
> In Select.js, check my fork
> https://github.com/bonzoSPb/react-select/blob/master/src/Select.js

Thank you @bonzoSPb ! 👍  It works great with a shadow DOM ! 💯 

@JedWatson Can you please consider the solution of @bonzoSPb ?

**Comment by syedaminx** [comment]
Can someone guide me as to what the `shadowRootNode` should be when using an iFrame?

**Comment by mikeyyyzhao** [comment]
I'm having a similar problem as @afvieira where react-select is showing the screen readers text. I've tried @bonzoSPb 's solution of wrapping my react-select around a custom MyNonceProvider class but the styling still looks wrong. Are there other solutions to this issue? It also looks like createEmotionCache and createEmotionCacheCustom aren't getting called in MyNonceProvider? 

```
import React, { Component } from 'react';
import createCache from '@emotion/cache';
import Select, { NonceProvider } from 'react-select';
import memoizeOne from 'memoize-one';

class MyNonceProvider extends NonceProvider {
  createEmotionCacheCustom(nonce) {
    return createCache({ nonce, key: 'custom-select-style', container: this.props.container });
  };

  createEmotionCache() {
    return memoizeOne(this.createEmotionCacheCustom);
  };
}

export default class CustomSelect extends Component {
  constructor(props) {
    super(props);
    this.state = {
    };

    console.log('MyNonceProvider', MyNonceProvider);
    console.log('NonceProvider', NonceProvider);
  }

  render() {
    const shadowRootNode = document.getElementById('iframe-container');
    return (
      <MyNonceProvider container={shadowRootNode}>
        <Select
          {...this.props}
        />
      </MyNonceProvider>
    );
  }
}

```

**Comment by mikeyyyzhao** [comment]
I'm running react-select in an iframe and getting the same styling bug as @afvieira where the screen reader gets displayed. I've tried running the code that @bonzoSPb and @norama have posted but the styling is still off. I tried console logging the nonce and seeing that it's returning null, could that be an issue?

```
import React, { Component } from 'react';
import createCache from '@emotion/cache';
import Select, { NonceProvider } from 'react-select';

class MyNonceProvider extends NonceProvider {
  createEmotionCache = (nonce) => {
    console.log('nonce', nonce);
    return createCache({ nonce, container: this.props.container });
  };
}

export default class MySelect extends Component {
  render() {
    let iframeContainer = document.getElementById('custom-iframe-container');

    return (
      <MyNonceProvider container={iframeContainer}>
        <Select
          {...this.props}
        />
      </MyNonceProvider>
    );
  }
};
```

**Comment by ebonow** [comment]
@mikeyyyzhao Would you kindly mind creating a codesandbox demonstrating this issue? It would be much easier to get eyes on this and get this pushed up in priority.

**Comment by mikeyyyzhao** [comment]
@ebonow sure thing, here's the [codesandbox](https://codesandbox.io/s/condescending-sun-h461k?file=/src/App.js). As you can see, react-select is showing the screen reader text as style is incorrect

**Comment by ebonow** [comment]
I'll try to review this and get back to you soon. I'm still fairly new at this CSS-in-JS / Emotion thing, but I'll see what I can find out as I do know this is affecting others.

**Comment by kopecmi8** [comment]
@JedWatson First of all thank you for great librarary!

Please consider on @bonzoSPb solution.  There is two separate problems when yours library is used in Shadow DOM. 
1. Problem with missing styles
2. Problem on mobile devices with touch screen

First problem can be easily fixed by adding custom NonceProvider and its not so annoying to fix it on users own.
 
The second is much worse. It cannot be fixed without modification of library code and its lead to dirty hacks when you need to use it. 

Thanks for your interest in.

**Comment by carlosribas** [comment]
This is an amazing library, but I am facing the same problem that @mikeyyyzhao described. I hope @ebonow finds a solution for that.

**Comment by ebonow** [comment]
Thanks for the ping @carlosribas .

I'm currently working on a PR for customized aria-live messages. While working on this, it occurred to me that `a11yText` should perhaps be its own component. Advantages of this includes:
- Customizable text with access to more props and state
- Add additional aria-live messaging without relying on prebaked implementation patterns/templates
- Ability to remove all aria-live messaging to give priority to other aria attributes
- Ability to apply style to the shadowDOM should the user want an unstyled Select or one that isn't reliant on Emotion

Regarding @bonzoSPb , I agree that we should add a composedPath check, though we'd likely want to do a null guard check. One improvement upon this bit of code though, would be a check inside the entire Menu and not just the menuListRef should incase someone want to add a Header or Footer to their MenuList. There's an existing issue logged for this, so was hoping to take it on in one go.

**Comment by carlosribas** [comment]
Sounds good. I'm really looking forward to this PR. Thank you @ebonow.

**Comment by sag1v** [comment]
> With my own NonceProvider, I am able to get the style sections injected in my webcomponent and react select works fine with it. But I could see duplicate style sections added every time I render react select component. Where I have 10 select elements, the styles are duplicated 10 time.
> 
> Is there a way to avoid this flood of style tags from emotion ? Thanks again

@pusulurm / @norama 

Did you manage to stop the duplication flood?  I'm facing the same issue when implementing @norama solution or even by creating my own NonceProvider from scratch.

**Comment by erkez** [comment]
I had this issue with styles when some of my application components were inserted into a shadow root.

I followed some of the suggestions here to simply use the emotion CacheProvider with a different container.

```typescript
import * as React from 'react';

import { CacheProvider } from '@emotion/core';
import createCache, { EmotionCache } from '@emotion/cache';

type Props = {
    stylesRoot?: HTMLElement;
    children: React.ReactNode;
};

export default function EmotionStyleSheetProvider(props: Props): React.ReactElement {
    const [cache, setCache] = React.useState<EmotionCache | null>(null);

    React.useEffect(() => {
        setCache(createCache({ container: props.stylesRoot }));
    }, [props.stylesRoot]);

    return (
        <React.Fragment>
            {cache != null && <CacheProvider value={cache}>{props.children}</CacheProvider>}
        </React.Fragment>
    );
}

```

In your application root, you could simply get the `ref` for a base div and pass it to a component similar to the one above. This actually managed to also solve issue #3309 for me.

Hope this helps.

**Comment by ptgamr** [comment]
@erkez solution works for us for a while.. but after this [PR!2361](https://github.com/emotion-js/emotion/pull/2361) of `@emotion/cache`, to fix a bug, make the solution no longer works..

Basically, the bug is that `@emotions/cache` modifies style tags 'belonging' to other cacheProviders... in this case, react-select has its own style provider (NonceProvider: https://github.com/JedWatson/react-select/blob/master/packages/react-select/src/NonceProvider.tsx)

```
<OurCustomerStyleProvider>
  <NonceProvider>
      ..........
  </NoneProvider>
</OurCustomerStyleProvider>
```

Because of the bug, `NoneProvider` will modify style tags belong to parent provider. (which is ours)

"The bug".. in this case, actually helping us... but it's now fixed, which mean we have to find another way.

**Comment by AndrewKuktenko** [comment]
Any solution here?

**Comment by myorkgitis** [comment]
> Any solution here?

I was able to get the styles to inject in production by hiding a `<Select>` component in my `_app.tsx`, outside of the page `Component`. 

My solution is here: https://github.com/JedWatson/react-select/issues/3309#issuecomment-883001216

**Comment by jfinnis** [comment]
yeah, here's the solution. Shit into your hand and force it down your face if you decided that @JedWatson had any solution to his library breaking every basic solution with web components. Sorry you trusted him. I'm for sure sorry we did. We have to select some completely different library, can't even afford time to research, bwcause he couldn't care enough to give a shit if his stuff worked with web compoments, the most common way people might interact with his library. It's been a fun week because I had the misfortune of dealing with @JedWatson

**Comment by jfinnis** [comment]
> yeah, here's the solution. Shit into your hand and force it down your face if you decided that @JedWatson had any solution to his library breaking every basic solution with web components. Sorry you trusted him. I'm for sure sorry we did. We have to select some completely different library, can't even afford time to research, bwcause he couldn't care enough to give a shit if his stuff worked with web compoments, the most common way people might interact with his library. It's been a fun week because I had the misfortune of dealing with @JedWatson. I pray you don't make the same mistake I did in trusting someone named @JedWatson. It would be a life-altering mistake to trust this library, we all deserve better than @JedWatson and I hope we all have better choices than @JedWatson .
> 
> Hey, @JedWatson prove me wrong! You won't. I know you won't. We all know you won't. It's embarrassing yet you're still here. This is so embarrassing for you. Can you just fucking delete yourself dude? You made a shitty library and didn't care how many people's lives you ruined.

Hello? @JedWatson is this thing on? Are you here with any control over any code you've ever released you scum?

**Comment by jfinnis** [comment]
Hey shithead no offense but hey wtf your shit is clearly broken is there any attempt to fix your shit or do we discard this entire library as completely fucking broken.

**Comment by jfinnis** [comment]
> @jfinnis u drunk or what? xD

No, but like, are you just nothing? Will you ever be just nothing? Because that's all you are now.

**Comment by Rall3n** [comment]
@jfinnis,

The behavior you are showing against a person you don't know and never met before is highly unacceptable.

We have reported your abusive/harassing behavior.

**Comment by Rall3n** [comment]
That being said:

I tested around a bit and got an example working.
https://codesandbox.io/s/react-custom-element-forked-ohlj0

The correct way to insert the styles into the shadow DOM is to use the `CacheProvider` component from `@emotion/react` in combination with a custom cache object created by `createCache` from `@emotion/cache` with the shadow root as its `container` attribute. Last but not least, the `Select` component is rendered as a child of `CacheProvider` that uses the created cache as its `value` prop.

```jsx

// Create shadow and mounting point
class XSelect extends HTMLElement {
  connectedCallback() {
    this.somehowCreateShadowAndMountPoint();
    
    const cache = createCache({
      container: this.shadowRoot,
      key: 'customElementCache'
    });
    
    ReactDOM.render(
      <CacheProvider value={cache}>
        <Select {...} />
      </CacheProvider>,
      this.mountPoint
    );
  }
}
```

It is important that the `container` attribute of the cache and the target element for `ReactDOM.render` must be different elements. If they are identical, there may be a race-condition between `emotion-js`, which tries to insert its stylesheets, and `react`, which tries to render the component, with the latter winning.

**Comment by jongbelegen** [comment]
@jfinnis, try the patch of @bonzoSPb, this works for me.

**Comment by jfinnis** [comment]
> @jfinnis, try the patch of @bonzoSPb, this works for me.

Thanks for pointing this out, this worked for me as well! This patch allows me to use the select dropdown on mobile: https://github.com/JedWatson/react-select/issues/3680#issuecomment-609679683

**Comment by baramik** [comment]
Was there any chance to apply the changes to make it work on mobile?

**Comment by CezaryDanielNowak** [comment]
Solutions here are pretty vague and troublesome. IMHO styles should be applied to the react root element, not to the page head by default.

**Comment by irhamputra** [comment]
IMHO it would be good that styles also should be provided in separated file, so people can add it manually where in their project

**Comment by forsureitsme** [comment]
So... won't fix and we'll have to add another dependency to our projects. Got it.

**Comment by DmitryMasley** [comment]
From my understanding, we could use NonceProvider. We only need to add a `container` prop to it in order to overwrite the default insertion point.

**Comment by Chrisdo82** [comment]
I also encountered this problem today and I don't get why this issue has been closed, sure there has been provided a solution, but it's merely a workaround and not a real fix.
